versioned_hdf5/tests/test_slicetools.py

```diff
@@ -47,6 +52,96 @@ def test_spaceid_to_slice(h5file):
                     assert_equal(a[s.raw], a[sel], f"{(start, count, stride, block)}")
 
 
+def test_build_slab_indices_and_offsets_dense(h5file):
+    chunks = (2, 3)
+    data = np.array(
+        [
+            [1, 2, 3, 4, 5],
+            [6, 7, 8, 9, 10],
+            [11, 12, 13, 14, 15],
+        ]
+    )
+    expect_raw_data = np.array(
+        [
+            # chunk (0, 0), offset 0
+            [1, 2, 3],
+            [6, 7, 8],
+            # chunk (0, 1), offset 2
+            [4, 5, 0],
+            [9, 10, 0],
+            # chunk (1, 0), offset 4
+            [11, 12, 13],
+            [0, 0, 0],
+            # chunk (1, 1), offset 6
+            [14, 15, 0],
+            [0, 0, 0],
+        ]
+    )
+    vf = VersionedHDF5File(h5file)
+    with vf.stage_version("r0") as sv:
+        sv.create_dataset("a", data=data, chunks=chunks)
+    virt_dset = h5file["_version_data/versions/r0/a"]
+    raw_data = h5file["_version_data/a/raw_data"]
+    np.testing.assert_array_equal(virt_dset[:], data, strict=True)
+    np.testing.assert_array_equal(raw_data[:], expect_raw_data, strict=True)
+
+    dcpl = virt_dset.id.get_create_plist()
+    indices, offsets = build_slab_indices_and_offsets(dcpl, data.shape, chunks)
+    np.testing.assert_array_equal(indices, [[1, 1], [1, 1]])
+    np.testing.assert_array_equal(offsets, [[0, 2], [4, 6]])
+
+
+def test_build_slab_indices_and_offsets_sparse(h5file):
+    chunks = (2, 3)
+    shape = (3, 5)
+    vf = VersionedHDF5File(h5file)
+    with vf.stage_version("r0") as sv:
+        sv.create_dataset(
+            "a", shape=shape, data=None, chunks=chunks, dtype=np.int64, fillvalue=123
+        )
+    virt_dset = h5file["_version_data/versions/r0/a"]
+    expect = np.full(shape, fill_value=123, dtype=np.int64)
+    np.testing.assert_array_equal(
+        virt_dset[:], np.full(shape, fill_value=123, dtype=np.int64), strict=True
+    )
+
+    dcpl = virt_dset.id.get_create_plist()
+    indices, offsets = build_slab_indices_and_offsets(dcpl, shape, chunks)
+    np.testing.assert_array_equal(indices, [[0, 0], [0, 0]])
+    np.testing.assert_array_equal(offsets, [[0, 0], [0, 0]])
+
+    # This test is recursive, as in order to create r1 on disk versioned_hdf5 is going
+    # to call build_slab_indices_and_offsets.
+    with vf.stage_version("r1") as sv:
+        dset = sv["a"]
+        dset[1, 3] = 456
+        dset[2, 0] = 789
+
+    virt_dset = h5file["_version_data/versions/r1/a"]
+    raw_data = h5file["_version_data/a/raw_data"]
+
+    expect[1, 3] = 456
+    expect[2, 0] = 789
+    np.testing.assert_array_equal(virt_dset[:], expect, strict=True)
+    expect_raw = np.array(
+        [
+            [123, 123, 123],
+            [456, 123, 123],
+            [789, 123, 123],
+            [123, 123, 123],
+            [123, 123, 123],  # FIXME Spurious extra chunk
```

Comment:
Is it weird that the spurious chunk in `r1` is at the end of the array?
Seems like there should be a chunk 0 (but not a chunk 3, since it duplicates chunk 0)

`r0` raw data:

```
chunk 0          chunk 1
[123, 123, 123], [123, 123, 123],
[123, 123, 123], [123, 123, 123],

chunk 2          chunk 3
[123, 123, 123], [123, 123, 123],
[123, 123, 123], [123, 123, 123],
```

`r1` raw data:
```

chunk 0          chunk 1
[123, 123, 123], [123, 123, 123],
[123, 123, 123], [456, 123, 123],

chunk 2          chunk 3
[789, 123, 123], [123, 123, 123],
[123, 123, 123], [123, 123, 123],
```

Reply:
Yes, it is very weird.

When you commit r0, there is a spurious chunk full of 123 at the beginning of the raw_data dataset.
It is spurious, because nothing in the data_dict points to it. `build_data_dict` raises `ValueError: Not a virtual storage layout (not a virtual storage layout)`. This is preempted by the special case logic from wrappers.py I linked above.

After writing r1, this spurious chunk moves to the *end* of raw_data.

Your diagram above isn't quite right. The actual one is

from the virtual dataset's POV:
```
r0
---
chunk 0 [0:2, 0:3]                 chunk 1 [0:2, 3:5]
not a virtual dataset              not a virtual dataset

chunk 2[2:3, 0:3]                  chunk 3 [2:3, 3:5]
not a virtual dataset              not a virtual dataset

r1
---
chunk 0 [0:2, 0:3]                 chunk 1 [0:2, 3:5]
not in get_virtual_count()         raw_data[0:2, :]
 
chunk 2[2:3, 0:3]                  chunk 3 [2:3, 3:5]         
raw_data[2:4, :]                   not in get_virtual_count()
```
from raw_data's POV:
```
r0
----
chunk 0 [0:2, :]
unused

r1
---
chunk 0 [0:2, :]
virtual_dataset[0:2, 3:5]

chunk 1 [2:4, :]
virtual_dataset[2:3, 0:3] 

chunk 2 [4:6, :]
unused
```